Return target_store to enable testing/logging stages after StoreToZarr

jbusecke opened the pull request:
Very much WIP. I am trying to return a singular element containing the store that was just written to. What I am currently trying here is producing multiple outputs (I guess one per fragment written?). 
@cisaacstern do you have any clever ideas to do this? I think it would open a host of functionality like testing and logging of successfully written stores.

Commit:
Return target_store to enable testing/logging stages after StoreToZarr

Commit:
Update transforms.py

Commit:
Update writers.py

cisaacstern commented on pangeo_forge_recipes/transforms.py:
```diff
@@ -468,4 +468,4 @@ def expand(self, datasets: beam.PCollection) -> beam.PCollection:
             target=self.get_full_target(), target_chunks=target_chunks
         )
-        return rechunked_datasets | StoreDatasetFragments(target_store=target_store)
+        return rechunked_datasets | StoreDatasetFragments(target_store=target_store) | beam.create([target_store])
```
```suggestion
        rechunked_datasets | StoreDatasetFragments(target_store=target_store)
        return target_store
```
@jbusecke, I think this would work?

Commit:
Update pangeo_forge_recipes/transforms.py

Co-authored-by: Charles Stern <62192187+cisaacstern@users.noreply.github.com>

jbusecke commented:
Hmm this did actually print nothing when running the recipe locally. I added a simple print stage at the end
```
| beam.Map(print)
```
which previously (as in my early commits here) did print the target store object (just not one, but many).
Do we need to wrap the target_store in a beam.create() or something else?

cisaacstern commented:
> Hmm this did actually print nothing when running the recipe locally. I added a simple print stage at the end

Not sure why it's not working for you. It does work for me. In my local clone, I put `beam.Map(print)` here:

https://github.com/jbusecke/pangeo-forge-recipes/blob/6922b8c4c6a4917b454bded2b0c14a1007780430/tests/test_end_to_end.py#L47-L53


```diff
            | StoreToZarr(
                target_root=tmp_target_url,
                store_name="store",
                target_chunks=target_chunks,
                combine_dims=pattern.combine_dim_keys,
            )
+           | beam.Map(print)
        )
```
Then running, e.g.

```
pytest -vxs tests/test_end_to_end.py -k "test_xarray_zarr[netcdf_local_file_pattern_sequential-target_chunks0-netcdf_local_paths_sequential_1d]" --disable-pytest-warnings
```
I get

```
====================================================== test session starts =======================================================
platform darwin -- Python 3.9.16, pytest-7.4.0, pluggy-1.2.0 -- /Users/charlesstern/miniconda3/envs/pfr-py39/bin/python3.9
cachedir: .pytest_cache
rootdir: /Users/charlesstern/Dropbox/pangeo/pangeo-forge-recipes
configfile: setup.cfg
plugins: nbmake-1.4.1, timeout-2.1.0, cov-4.1.0, lazy-fixture-0.6.3, anyio-3.7.0
timeout: 30.0s
timeout method: signal
timeout func_only: False
collected 22 items / 21 deselected / 1 selected                                                                                  

tests/test_end_to_end.py::test_xarray_zarr[netcdf_local_file_pattern_sequential-target_chunks0-netcdf_local_paths_sequential_1d] <zarr.storage.FSStore object at 0x19e014550>
```
Note the `<zarr.storage.FSStore object at 0x19e014550>` printed at the end of the last line (scroll right).

cisaacstern commented:
Also, can you split this into its own PR against `pangeo-forge/pangeo-forge-recipes@main` ?

Having it on top of dynamic chunking will make this harder to review, IMO.

As noted in https://github.com/pangeo-forge/pangeo-forge-recipes/issues/559#issuecomment-1671800216, you could merge this with `dynamic_chunks_2` into a separate `dev` branch on your fork, if you want the combined changes available for your development work. But having them layered together makes it harder to evaluate each as a PR to `main`, if that makes sense.

jbusecke commented:
Makes total sense, I basically made these to drive very experimental work over at https://github.com/leap-stc/cmip6-leap-feedstock.

jbusecke commented:
> > Hmm this did actually print nothing when running the recipe locally. I added a simple print stage at the end
> 
> Not sure why it's not working for you. It does work for me. In my local clone, I put `beam.Map(print)` here:

I assume then it is a problem with using pgf-runner locally at my end...I tried to adjust the `requirements.txt` but I am not sure that this even is relevant for a local bakery run?

jbusecke commented:
Indeed this seems to work. I was able to implement a rudimentary test stage like this:

```python
@dataclass
class TestDataset(beam.PTransform):
    """
    Test stage for data written to zarr store
    """
    iid: str
    
    def _check_and_print_store(self,store: zarr.storage.FSStore) -> zarr.storage.FSStore:
        iid_schema = "mip_era.activity_id.institution_id.source_id.experiment_id.variant_label.table_id.variable_id.grid_label.version"
        import xarray as xr
        import logging
        logger = logging.getLogger()
        ds = xr.open_dataset(store, engine='zarr', chunks={})
        print(ds)
        logger.info(f"Reloaded {ds =}")
        # assert (ds.time.diff('time') > 0).all()
        
        for facet_value, facet in zip(self.iid.split('.'), iid_schema.split('.')):
            if not 'version' in facet:
                print(f"Checking {facet = } in dataset attributes")
                assert ds.attrs[facet] == facet_value
          
        return store
    
    def expand(self, pcoll: beam.PCollection) -> beam.PCollection:
        return (pcoll
            | "Testing Xarray Dataset" >> beam.Map(self._check_and_print_store)
        )

recipes = {}
for iid, urls in url_dict.items():
    pattern = pattern_from_file_sequence(
        urls,
        concat_dim='time'
        )
    recipes[iid] = (
        beam.Create(pattern.items())
        | OpenURLWithFSSpec()
        | OpenWithXarray(xarray_open_kwargs={"use_cftime":True}) # do not specify file type to accomodate both ncdf3 and ncdf4
        | Preprocessor(urls=urls)
        | StoreToZarr(
            store_name=f"{iid}.zarr",
            combine_dims=pattern.combine_dim_keys,
            target_chunks={'time':400},
            # target_chunk_size='150MB',
            # target_chunks_aspect_ratio = target_chunks_aspect_ratio,
            # size_tolerance=0.4
            )
        | TestDataset(iid=iid)
        )
```

...and I am getting this in the logal pgf-runner output: 

```
...
Worker: severity: INFO timestamp {   seconds: 1691681703   nanos: 893973827 } message: "Python sdk harness exiting." log_location: "/srv/conda/envs/cmip6-leap-feedstock/lib/python3.9/site-packages/apache_beam/runners/worker/sdk_worker_main.py:184" thread: "MainThread" 
<xarray.Dataset>
Dimensions:    (lat: 144, bnds: 2, lon: 192, time: 52920)
Coordinates:
  * lat        (lat) float64 0.0 0.0 0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0 0.0 0.0
    lat_bnds   (lat, bnds) float64 dask.array<chunksize=(144, 2), meta=np.ndarray>
  * lon        (lon) float64 0.0 0.0 0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0 0.0 0.0
    lon_bnds   (lon, bnds) float64 dask.array<chunksize=(192, 2), meta=np.ndarray>
  * time       (time) object 1850-01-01 00:00:00 ... 1850-01-01 00:00:00
    time_bnds  (time, bnds) object dask.array<chunksize=(400, 2), meta=np.ndarray>
Dimensions without coordinates: bnds
Data variables:
    psl        (time, lat, lon) float32 dask.array<chunksize=(400, 144, 192), meta=np.ndarray>
Attributes: (12/44)
    Conventions:            CF-1.7 CMIP-6.2
    activity_id:            CMIP
    branch_method:          standard
    branch_time_in_child:   0.0
    branch_time_in_parent:  0.0
    cmor_version:           3.4.0
    ...                     ...
    table_id:               day
    table_info:             Creation Date:(13 December 2018) MD5:f0588f7f55b5...
    title:                  UKESM1-0-LL output prepared for CMIP6
    variable_id:            psl
    variable_name:          psl
    variant_label:          r14i1p1f2
Checking facet = 'mip_era' in dataset attributes
Checking facet = 'activity_id' in dataset attributes
Checking facet = 'institution_id' in dataset attributes
Checking facet = 'source_id' in dataset attributes
Checking facet = 'experiment_id' in dataset attributes
Checking facet = 'variant_label' in dataset attributes
Checking facet = 'table_id' in dataset attributes
Checking facet = 'variable_id' in dataset attributes
Checking facet = 'grid_label' in dataset attributes
...
```

This is super cool! 

@cisaacstern do you think we need more tests for this, or could we merge this as is?

jbusecke commented:
I have moved this to https://github.com/pangeo-forge/pangeo-forge-recipes/pull/562, since apparently I am too dense to move the target fork/repo on this PR 🙄.
